### DISCUSSION TITLE
How can i include a custom ios framework inside the ios library created by the tool?

### DISCUSSION THREAD

**Discussion by angelos3lex** [callstack/react-native-builder-bob#379] [Q&A] [answered]
I have a custom ios framework (locally, private) which i include in a native ios app. It contains some ui components.

I want the same framework, to export it to a react native library. 

So, i created one with this tool, but i can not figure out how i can include my framework inside the ios xcodeproj so that i can import in the ViewManager.swift file.

Any tutorial somewhere?

**Comment by atlj** [comment]
Hey, it's possible. You need to edit your `.podspec` file found on the project's root. Please put this line:

```ruby
  # Specify path to the framework
  s.vendored_frameworks = "path/to/your_framework.xcframework"
```

After you install the pods again, you can import the framework to your project.

**Reply by angelos3lex** [reply]
Hey Thanks for your help!

I have a `.framework` instead of a `.xcframework`. 
I added it as you suggested: 
`s.vendored_frameworks = "/Users/my_user/.../my_framework.framework"`

`pod install` ran successfully.
But i can't see how i can use it inside the library.

I added `import my_framework` inside the ViewManager, but during build i get: `No such module 'my_framework'`

**Reply by angelos3lex** [reply]
Edit on above answer, i mistakenly added the `s.vendored_frameworks` inside the `if` about `newArchitectureEnabled`, so it was ignored. I moved it above, and i got error that i can't have absolute paths. 

So i made it relative, ran `yarn` successfully which ran `pod install`, but still the problem persists. I get `no such module my_framework` in xCode. Maybe there is any extra missing step?

**Reply by angelos3lex** [reply, marked as answer]
After all i had success! I had to copy the `my_framework.framework` on the same parent folder as the `.podspec` file.
Inside `.podspec` i had to import it like this:
`s.vendored_frameworks = "my_framework.framework"`

Thanks @atlj for your hint & help! I think i understand a bit better how all these stuff are linked!

**Reply by atlj** [reply]
I'm glad it solved your problem. Marked your answer as solution. Have a great day 🚀.

**Comment by angelos3lex** [comment]
Hey @atlj here i come again!
Everything works on my end, cause i build the `.framework` and the react-native app on same mac.

Though, on another mac, if i try to build the react-native app, i get error saying that the `.framework` was compiled by a different compiler version:
```
library/ios/LibraryView.swift:1:8: Compiled module was created by a different version of the compiler; rebuild 'framework-name' and try again: library/framework-name.framework/Modules/framework-name.swiftmodule/x86_64-apple-ios-simulator.swiftmodule
```
That's very weird because when i build the `.framework` i have chosen under architecture the `$(ARCHS_STANDARD)` and i have set `Build Active Architecture only` to `No`. Also, i see that the produced `.framework` contains all arm64 and x86_64 swiftmodules.

Any ideas? That's a blocker if i want to share my library with other people, as they will get the above error during build.

**Reply by atlj** [reply]
Could you please also set "Build Library For Distribution" to "Yes" in the build settings when creating the framework? I'm not sure if you have access to the source code of the framework file you mentioned, but they should provide you with such a framework. Otherwise, everyone who uses your library will need to have the same version of Xcode installed. 
https://developer.apple.com/documentation/xcode/creating-a-multi-platform-binary-framework-bundle#Set-up-the-project

**Reply by angelos3lex** [reply]
Thanks for the hint! I'm building the framework using above configuration in xcode v14.2, and then, when trying to build react native app on another mac which has xcode v14.3, i get following error:
```
Failed to build module 'library'; this SDK is not supported by the compiler (the SDK is built with 'Apple Swift version 5.7.2 (swiftlang-5.7.2.135.5 clang-1400.0.29.51)', while this compiler is 'Apple Swift version 5.8 (swiftlang-5.8.0.124.2 clang-1403.0.22.11.100)'). Please select a toolchain which matches the SDK.
```
Respectively if i build the framework in xcode v14.3 and try to build react native app in xcode v14.2, i get same error with swift versions changed accordingly.

So even with "Build Library For Distribution" set to "Yes", seems like having the same xcode version as the one that the framework is created is essential?!

**Reply by atlj** [reply]
Yes it's entirely dependent on how the framework was originally built. If you are able to contact the owners of the SDK, you can ask them to enable this option and build a new version. This is actually not a react native related bug. It's more of an iOS one. If you were to try using the SDK on a native iOS app, you would most likely get the same result.

**Reply by angelos3lex** [reply]
@atlj I'm the owner of the sdk, and i built it with this option enabled, though i get the above error about swift versions mismatch.
After searching i ended up on the `.xcframework` format, since that's the solution to distribute a framework which is playable across architectures. I successfully built one, but when i build the react native app which uses this xcframework, i face another error:
-> `cannot load underlying module MapboxMaps`
My podspec is:
```
 s.dependency 'MapboxMaps', '10.14.0'
 s.vendored_frameworks = "myframework.xcframework"
```
So, my framework depends on MapboxMaps, and even though building it as a .framework and importing on same machine works in react native app, when trying to export as xcframework and let others use it, i face above error. 

At this point, I am trying to find a lifeline and I was wondering if you would be interested to work on this, by payment of course.

**Reply by atlj** [reply]
Could you also check this answer? https://developer.apple.com/forums/thread/668420
